cpp/serve/engine_actions/batch_draft.cc

```diff
@@ -0,0 +1,213 @@
+/*!
+ *  Copyright (c) 2023 by Contributors
+ * \file serve/engine_actions/batch_spec_decode.cc
+ */
+
+#include "../config.h"
+#include "../model.h"
+#include "../sampler.h"
+#include "action.h"
+#include "action_commons.h"
+
+namespace mlc {
+namespace llm {
+namespace serve {
+
+/*!
+ * \brief The action that runs draft proposal for requests in the
+ * `running_queue` of engine state. Preempt low-priority requests
+ * accordingly when it is impossible to decode all the running requests.
+ */
+class BatchDraftActionObj : public EngineActionObj {
+ public:
+  explicit BatchDraftActionObj(Array<Model> models, Sampler sampler,
+                               Optional<EventTraceRecorder> trace_recorder, int draft_length = 4)
+      : models_(std::move(models)),
+        sampler_(std::move(sampler)),
+        trace_recorder_(std::move(trace_recorder)),
+        draft_length_(draft_length) {}
+
+  Array<Request> Step(EngineState estate) final {
+    // - Only run spec decode when there are two models (llm+ssm) and >=1 running requests.
+    if (models_.size() != 2 || estate->running_queue.empty()) {
+      return {};
+    }
+
+    // Preempt requests when decode cannot apply.
+    while (!CanDecode(estate->running_queue.size())) {
+      PreemptLastRunningRequest(estate);
+    }
+
+    auto tstart = std::chrono::high_resolution_clock::now();
+
+    // NOTE: Right now we only support decode all the running requests at a time.
+    int num_requests = estate->running_queue.size();
+
+    // The first model doesn't get involved in draft proposal.
+    for (int model_id = 1; model_id < models_.size(); ++model_id) {
+      // draft_length_ rounds of draft proposal.
+      for (int draft_id = 0; draft_id < draft_length_; ++draft_id) {
+        // Collect
+        // - the last committed token,
+        // - the request states,
+        // - the sampling parameters,
+        // of each request.
+        std::vector<int> input_tokens;
+        Array<String> request_ids;
+        std::vector<int64_t> request_internal_ids;
+        Array<RequestModelState> mstates;
+        Array<GenerationConfig> generation_cfg;
+        input_tokens.reserve(num_requests);
+        request_ids.reserve(num_requests);
+        request_internal_ids.reserve(num_requests);
+        mstates.reserve(num_requests);
+        generation_cfg.reserve(num_requests);
```

Comment:
Only `input_tokens` is collected each time, and the other 4 arrays are shared across different draft_ids right? I think we collect the four first outside and do not need to collect them repeatedly every time.